pkg/apply/task/apply_task.go

```diff
@@ -98,7 +109,7 @@ func (a *ApplyTask) Start(taskContext *taskrunner.TaskContext) {
 
 		// Create a new instance of the applyOptions interface and use it
 		// to apply the objects.
-		ao, err := applyOptionsFactoryFunc(taskContext.EventChannel(),
+		ao, dynamic, err := applyOptionsFactoryFunc(taskContext.EventChannel(),
```

Comment:
Maybe we should do this in two steps, so we first create the dynamic client and then pass that into `applyOptionsFactoryFunc`?

Reply:
The `dynamic` is created from `util.Factory`, which is passed into `applyOptionsFactoryFunc`, but `dynamic` is not the only one needed to create an `ApplyOptions`. It needs a `discovery` client as well. Both `dynamic` and `discovery` are from `util.Factory`. It's better to put the code for getting both `dynamic` and `discovery` together.

Another approach is to passed both `dynamic` and `discovery` to the ApplyTask. Then the `util.Factory` is not needed. This may need a bit refactoring on current code and test. If you prefer this approach, we can do it in a different PR.